cwl/src/main/scala/cwl/InitialWorkDirRequirement.scala

```diff
@@ -0,0 +1,70 @@
+package cwl
+
+import cwl.InitialWorkDirRequirement.{ExpressionOrString, IwdrListingArrayEntry, _}
+import eu.timepit.refined.W
+import shapeless.{:+:, CNil, _}
+
+final case class InitialWorkDirRequirement(
+                                            `class`: W.`"InitialWorkDirRequirement"`.T,
+                                            listing: Array[IwdrListingArrayEntry] :+: ExpressionOrString :+: CNil
+                                          ) {
+  val listings: Array[IwdrListingArrayEntry] = listing match {
+    case IwdrListingArray(array) => array
+    case ExpressionOrString(eos) => Array(Coproduct[IwdrListingArrayEntry](eos))
+  }
+
+  override def toString: WorkflowStepInputId =
+    s"""InitialWorkDirRequirement(
+      |  ${listings.mkString(System.lineSeparator + "  ")}
+      |)""".stripMargin
+}
+
+/**
+  *  Short for "Directory Entry"
+  *  @see <a href="http://www.commonwl.org/v1.0/CommandLineTool.html#Dirent">Dirent Specification</a>
+  *
+  *  Split into two cases because entryName is only optional if entry is an Expression
+  */
+trait Dirent {
+  def writable: Option[Boolean]
+  def writableWithDefault = writable.getOrElse(false)
+}
+
+final case class ExpressionDirent(
+                                   entry: Expression,
+                                   entryname: Option[ExpressionOrString],
+                                   writable: Option[Boolean]
+                                 ) extends Dirent
+
+final case class StringDirent(
+                               entry: String,
+                               entryname: ExpressionOrString,
+                               writable: Option[Boolean]
+                             ) extends Dirent
+
+object InitialWorkDirRequirement {
+
+  final type IwdrListingArrayEntry = File :+: Directory :+: StringDirent :+: ExpressionDirent :+: ExpressionOrString :+: CNil
+
+  object IwdrListingArrayEntry {
+    object StringDirent {
+      def unapply(e: IwdrListingArrayEntry): Option[(String, ExpressionOrString, Boolean)] =
+        e.select[StringDirent].map(sd => (sd.entry, sd.entryname, sd.writableWithDefault))
```

Comment:
I see here the effort to use writableWithDefault 👍 ( but I still wonder if it could be enforced via the type system).